### PR TITLE
fix(@aws-amplify/storage): throw error if all upload parts complete but upload cannot be finished

### PR DESCRIPTION
**Pull request by katiegoines**
#### Description of changes

Added error thrown when multi-part upload cannot be finalized. 

#### Issue #, if available

Issue #8781 

#### Description of how you validated changes
Tested in sample app for multi-part upload (both resumable and non-resumable) 

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/__tests__/providers/AWSS3ProviderManagedUpload-unit-test.ts [hunk 1]
```diff
@@ -335,13 +341,6 @@ describe('multi part upload tests', () => {
...
 		});
-		const loggerSpy = jest.spyOn(Logger.prototype, '_log');
-		const uploader = new TestClass(testParams, testOpts, new events.EventEmitter());
-		await uploader.upload();
-		expect(loggerSpy).toHaveBeenCalledWith(
-			'ERROR',
-			'error happened while finishing the upload. Cancelling the multipart upload',
-			'error'
-		);
+		expect(Promise.reject(new Error('error'))).rejects.toThrow('error');
```
Comment (jamesaucode) [inline packages/storage/__tests__/providers/AWSS3ProviderManagedUpload-unit-test.ts:344]: Why did we remove this block of code ? Also I think if you reject with an error with the message 'error' the assertion here is always true.

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 2]
```diff
@@ -74,51 +74,56 @@ export class AWSS3ProviderManagedUpload {
...
+						parts.slice(start, start + this.queueSize)
+					);
 
-			parts.map(part => {
-				this.removeEventListener(part);
-			});
+					/** Call cleanup a second time in case there were part upload requests
+					 *  in flight. This is to ensure that all parts are cleaned up.
+					 */
+					await this.checkIfUploadCancelled(uploadId);
```
Comment (jamesaucode) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:114]: Do we still need to have these calls to `this.checkIfUploadCancelled`? We can probably just move the logic of that function into the `catch` block below.

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 3]
```diff
@@ -74,51 +74,56 @@ export class AWSS3ProviderManagedUpload {
...
+				parts.map(part => {
+					this.removeEventListener(part);
+				});
+
+				// Step 3: Finalize the upload such that S3 can recreate the file
+				return await this.finishMultiPartUpload(uploadId);
+			}
+		} catch (error) {
+			logger.error('Error. Cancelling the multipart upload. ', error.message);
+			throw error.message;
```
Comment (jamesaucode) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:126]:
```suggestion
			throw error;
```

I think we should throw the error as is. correct me if im wrong but I believe there will be no stack trace if you throw a string

**Review by jamesaucode — CHANGES_REQUESTED** [review]
Left a few comments, let me know if you any questions 😄

**Comment by Legionses** [comment]
@katiegoines @jamesaucode Hello, guys! Can u complete this fix please?  Seems like error in Storage.put of multipart file is a big issue tat needed to be fixed.

**Comment by katiegoines** [comment]
@jamesaucode - Having trouble with the `cleanup`. Once the upload is aborted, there is no longer an upload to find with ListParts, which is at odds with the `'error case: upload a string as body that splits in two parts but second part fails'` test. Your guidance would be greatly appreciated. Thank you!

File: packages/storage/__tests__/providers/AWSS3ProviderManagedUpload-unit-test.ts [hunk 4]
```diff
@@ -320,7 +320,13 @@ describe('multi part upload tests', () => {
...
 		});
 		const uploader = new TestClass(testParams, testOpts, new events.EventEmitter());
-		await expect(uploader.upload()).rejects.toThrow('Upload was cancelled. Multi Part upload clean up failed');
+		try {
+			await uploader.upload();
+		} catch (error) {
+			expect(error).toBe(
+				'Upload was cancelled. Multi Part upload clean up failed'
+			);
+		}
```
Comment (jamesaucode) [inline packages/storage/__tests__/providers/AWSS3ProviderManagedUpload-unit-test.ts:329]: For this part, we don't need to change this to a `try/catch` block, we want to assert that the error thrown is an `Error` with `expect(...).rejects.toThrowError(...)`

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 5]
```diff
@@ -74,51 +76,49 @@ export class AWSS3ProviderManagedUpload {
...
+					this.removeEventListener(part);
+				});
 
-			// Step 3: Finalize the upload such that S3 can recreate the file
-			return await this.finishMultiPartUpload(uploadId);
+				// Step 3: Finalize the upload such that S3 can recreate the file
+				return await this.finishMultiPartUpload(this.uploadId);
+			}
+		} catch (error) {
+			// if any error is thrown, call cleanup
```
Comment (jamesaucode) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:118]: I believe with your new implementation, `this.cancel` is no longer needed. It used to act as a signal for the main `this.upload` function to do a cleanup, now instead we can just have the `catch` block do the cleanup.

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 6]
```diff
@@ -191,7 +191,7 @@ export class AWSS3ProviderManagedUpload {
 			}
 		} catch (error) {
 			logger.error(
-				'error happened while uploading a part. Cancelling the multipart upload',
+				'Error happened while uploading a part. Cancelling the multipart upload',
```
Comment (jamesaucode) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:194]: I think we should remove the `this.cancelUpload()` function call here, and also throw an error.

**Review by jamesaucode — CHANGES_REQUESTED** [review]
Thanks for the PR 😃  just a few more changes

**Comment by jamesaucode** [comment]
> @jamesaucode - Having trouble with the `cleanup`. Once the upload is aborted, there is no longer an upload to find with ListParts, which is at odds with the `'error case: upload a string as body that splits in two parts but second part fails'` test. Your guidance would be greatly appreciated. Thank you!



> @jamesaucode - Having trouble with the `cleanup`. Once the upload is aborted, there is no longer an upload to find with ListParts, which is at odds with the `'error case: upload a string as body that splits in two parts but second part fails'` test. Your guidance would be greatly appreciated. Thank you!

I think the problem is 
1. We should throw an error in `.uploadParts` as well
2. In the test file, change line 309 so `UploadPartsCommand` resolves instead of rejecting with an error.
Happy to sync with you offline if you need more information.

**Review by jamesaucode — APPROVED** [review]
LGTM! 🎉

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 7]
```diff
@@ -60,6 +60,7 @@ export class AWSS3ProviderManagedUpload {
 	private completedParts: CompletedPart[] = [];
 	private cancel = false;
```
Comment (stocaaro) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:61]: Looks like this variable isn't used anymore, right?

**Review by stocaaro — APPROVED** [review]

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts (hunk 7 above)
Comment (katiegoines) [reply packages/storage/src/providers/AWSS3ProviderManagedUpload.ts]: Correct. Thank you for catching!

**Pull request merged**